feat: add audit log for assessment mode

rschlaefli opened the pull request:
<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

* **New Features**
  * Added a standalone Audit Service with HTTP endpoints, readiness & metrics, and integrated audit logging across backend, workers, auth, and frontend assessment flows.
* **Chores/Infrastructure**
  * Added Azurite (Azure Storage emulator) to local/CI, startup readiness checks, Docker/Helm manifests, env examples, and release workflows; standardized Postgres test images.
* **Tests**
  * Added extensive API, integration, performance, and scenario test suites and CI jobs to build, test, and publish the Audit Service.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

<!-- AI_TRIAGE_CLICKUP_LINK_START -->
## ClickUp Links

> *This was generated by AI during triage.*

- Primary task: [MeF audit logging and pure-infra semester-start readiness](https://app.clickup.com/t/86caazaea)
<!-- AI_TRIAGE_CLICKUP_LINK_END -->

rschlaefli committed:
draft for audit log

rschlaefli committed:
cleanup

rschlaefli committed:
README and testing for audit-log

rschlaefli committed:
initial deployment config

rschlaefli committed:
test rework

rschlaefli committed:
improve docs and example files

rschlaefli committed:
Merge branch 'v3-assessment' of https://github.com/uzh-bf/klicker-uzh into audit-log

rschlaefli committed:
clients for audit logging

rschlaefli committed:
WIP

rschlaefli committed:
Merge branch 'v3-assessment' of https://github.com/uzh-bf/klicker-uzh into audit-log

rschlaefli committed:
wip

rschlaefli committed:
Merge branch 'v3-assessment' of https://github.com/uzh-bf/klicker-uzh into audit-log

# Conflicts:
#	packages/graphql/src/services/accounts.ts
#	pnpm-lock.yaml

rschlaefli committed:
remove tenantId, add test scripts

rschlaefli committed:
docs build separately

rschlaefli committed:
Merge branch 'v3-assessment' of https://github.com/uzh-bf/klicker-uzh into audit-log

# Conflicts:
#	_run_app_dependencies.sh
#	_run_app_dependencies_macos.sh
#	_run_app_dependencies_wsl.sh
#	apps/backend-docker/.env.example
#	pnpm-lock.yaml
#	project/plans_archive/PLAN-audit-log-clients.md
#	project/plans_archive/PLAN-audit-log.md
#	project/plans_archive/PROJECT-audit-log.md

rschlaefli committed:
update tests, queue planning

rschlaefli committed:
Merge branch 'v3-assessment' of https://github.com/uzh-bf/klicker-uzh into audit-log

# Conflicts:
#	pnpm-lock.yaml

rschlaefli committed:
Merge branch 'v3-assessment' of https://github.com/uzh-bf/klicker-uzh into audit-log

# Conflicts:
#	.github/workflows/cypress-testing.yml
#	apps/backend-docker/.env.cypress
#	apps/backend-docker/.env.example
#	packages/util/src/index.ts
#	pnpm-lock.yaml

rschlaefli committed:
Merge branch 'v3-assessment' of https://github.com/uzh-bf/klicker-uzh into audit-log

rschlaefli committed:
Merge branch 'v3' of https://github.com/uzh-bf/klicker-uzh into audit-log

# Conflicts:
#	.versionrc.js
#	_run_app_dependencies.sh
#	pnpm-lock.yaml

rschlaefli commented on .github/workflows/cypress-testing.yml:
```diff
@@ -330,6 +330,19 @@ jobs:
           SERVER_LOGGER_LEVEL: debug
           SERVER_LOGGER_FORMAT: console
 
+      azurite:
```
Azurite is a local emulator for blob/table/queue storage. We could probably start it only in assessment mode though (maybe another flag for the run app dependencies?)

rschlaefli commented on deploy/env-prod-v3/values.yaml:
```diff
@@ -533,6 +533,29 @@ responseApi:
+  azureStorage:
+    tableName: 'auditlogs-prod'
+
+  corsOrigins: 'https://assessment.klicker.uzh.ch'
```
do we ever need any frontend audit logging from manage?

rschlaefli committed:
small issues

rschlaefli committed:
add docker builds and test workflow

rschlaefli committed:
simplification and cleanup, single schema

rschlaefli committed:
fix ts issues

rschlaefli committed:
add azurite to cypress

rschlaefli committed:
Merge branch 'v3' of https://github.com/uzh-bf/klicker-uzh into audit-log

# Conflicts:
#	.github/scripts/wait-for-services.sh

rschlaefli committed:
remove options from azurite service

rschlaefli committed:
chore: add invitations file for dev_assessment

rschlaefli committed:
improvements and initial audit logging through context

rschlaefli committed:
further improvements

rschlaefli committed:
Merge branch 'v3' of https://github.com/uzh-bf/klicker-uzh into audit-log

# Conflicts:
#	packages/types/src/index.ts

rschlaefli committed:
Merge branch 'v3' of https://github.com/uzh-bf/klicker-uzh into audit-log

rschlaefli committed:
audit client usage in response api and processor

rschlaefli committed:
further improvements

rschlaefli committed:
lockfile update

rschlaefli committed:
use audit client to submit in frontend

rschlaefli committed:
response submission logged, cypress fixes

rschlaefli committed:
further work

rschlaefli committed:
update env files

rschlaefli committed:
apps audit in audit test workflow

rschlaefli committed:
remove script

rschlaefli committed:
reject unauthorized false in dev mode

rschlaefli committed:
Merge branch 'v3' of https://github.com/uzh-bf/klicker-uzh into audit-log

rschlaefli committed:
more env files and consistency updates

rschlaefli committed:
update test workflows

rschlaefli committed:
update plan

rschlaefli committed:
start the audit app in the background

rschlaefli committed:
env for audit service startup

rschlaefli committed:
extend allowed public actions, use AuditAction in test

rschlaefli committed:
log successful login for participant via auth

rschlaefli committed:
new user control types

rschlaefli committed:
live quiz audits

rschlaefli committed:
plan reorg

rschlaefli committed:
audit log in import script

rschlaefli committed:
Merge branch 'v3' of https://github.com/uzh-bf/klicker-uzh into audit-log

# Conflicts:
#	.gitignore
#	apps/auth/package.json
#	pnpm-lock.yaml

rschlaefli committed:
fix: formatting issues

rschlaefli committed:
cleanup

rschlaefli committed:
add app origin to test audit

rschlaefli committed:
fix: participant cookie renamed

rschlaefli committed:
dont accept temp participants

rschlaefli committed:
wait for services in util

rschlaefli commented on .github/workflows/cypress-testing.yml:
```diff
@@ -44,7 +44,7 @@ jobs:
     services:
       postgres:
-        image: postgres:15
+        image: postgres:14.18
```
made this actually aligned with what we have on azure

rschlaefli commented on .github/workflows/v3_audit-prod.yml:
```diff
@@ -0,0 +1,47 @@
+name: Build Docker image for audit (Production)
```
@jabbadizzleCode will require new pipelines on the your branch as well

rschlaefli commented on apps/audit/src/middleware/auth.ts:
```diff
@@ -0,0 +1,71 @@
+  role: string
+}
+
+export async function authMiddleware(
```
backend needs to send a fixed token in the header. frontend sends the student login JWT. could probably merge to both being a JWT with different issuers and according validation, but for now this will be enough

rschlaefli commented on apps/audit/src/routes/audit-private.ts:
```diff
@@ -0,0 +1,126 @@
+      const startTime = Date.now()
+      const event = c.req.valid('json')
+
+      metrics.requestsTotal.inc()
```
idea: expose certain metrics for "prometheus", which will run on the new infra and can monitor apps. it could scale up the replicas if e.g. we have a lot of total requests, or many more requests than writes, etc.

if this works well, should expand on this in our own backend too, so we dont have only CPU/GPU usage metrics

rschlaefli commented on apps/audit/src/routes/audit-private.ts:
```diff
@@ -0,0 +1,126 @@
+  tableClient: AuditTableClient
+}
+
+export function registerAuditPrivateRoutes(
```
the actual endpoints are very simple on purpose. we dont reject events as long as they are properly authenticated (not a lot of validation on e.g. allowed types beside the public endpoint). to make sure that w log something invalid rather than dropping it with an error message

rschlaefli commented on apps/audit/src/routes/audit-public.ts:
```diff
@@ -0,0 +1,252 @@
+      }
+
+      const rateLimitKey = `public:${participant.participantId}`
+      if (!checkRateLimit(rateLimitKey, 100, 60000)) {
```
we have to tune this, but a given participant should not be able to send too many events to the public endpoint (just for abuse prevention). maybe 100 in one minute is too low though. we could also have rate limits on the public endpoint before validation (e.g., based on IP).

rschlaefli commented on apps/audit/src/routes/audit-public.ts:
```diff
@@ -0,0 +1,252 @@
+
+      const {
+        attributes,
+        subject: _subjectIgnored,
```
in the public case, the subject is always the user that sends to the public endpoint, so we ignore what they might pass along in the request

rschlaefli commented on apps/audit/src/routes/audit-public.ts:
```diff
@@ -0,0 +1,252 @@
+        ...rest
+      } = (rawBody ?? {}) as Record<string, unknown>
+
+      const attributeRecord =
```
attributes could be whatever is passed in the request right now, we might need to restrict this in the future. zod does some validation on the max size of it already (as table storage caps it as well)

rschlaefli commented on apps/audit/src/schemas/audit-event.ts:
```diff
@@ -0,0 +1,119 @@
+  AuditAction.PARTICIPANT_JOIN_QUIZ,
+  AuditAction.PARTICIPANT_QUIZ_PIN_SUCCESS,
+  AuditAction.PARTICIPANT_QUIZ_PIN_FAILED,
+  AuditAction.CLIENT_ERROR,
```
any error on the client caught by an error boundary

rschlaefli commented on apps/audit/src/schemas/audit-event.ts:
```diff
@@ -0,0 +1,119 @@
+import { AuditAction, AuditScope } from '@klicker-uzh/types'
+import { z } from 'zod'
+
+export const ALLOWED_PUBLIC_ACTIONS = [
```
implementation in the UI is WIP on the other branch

rschlaefli commented on apps/audit/src/schemas/audit-event.ts:
```diff
@@ -0,0 +1,119 @@
+
+const ALLOWED_PUBLIC_ACTIONS_SET = new Set<string>(ALLOWED_PUBLIC_ACTIONS)
+
+const CORRELATION_ID_REGEX = /^[a-f0-9]{32}$/i
```
alignment with our real correlation ID is WIP on the other branch

rschlaefli commented on apps/audit/src/storage/entities.ts:
```diff
@@ -0,0 +1,100 @@
+export interface AuditTableEntity {
+  partitionKey: string
+  rowKey: string
+  scope: string
```
where does the log come from

public (any ui) | internal (backends) | worker (hatchet/async processing)

sjschlapbach replied:
@rschlaefli If we know that the value will be one of three strings, is there a reason, why we do not type this accordingly?

rschlaefli commented on apps/audit/src/storage/entities.ts:
```diff
@@ -0,0 +1,100 @@
+  partitionKey: string
+  rowKey: string
+  scope: string
+  subject: string
```
the user/entity that caused/triggered the event

rschlaefli commented on apps/audit/src/storage/entities.ts:
```diff
@@ -0,0 +1,100 @@
+  rowKey: string
+  scope: string
+  subject: string
+  action: string
```
what the action of the event was (e.g., a response was submitted)

always used via constants

rschlaefli commented on apps/audit/src/storage/entities.ts:
```diff
@@ -0,0 +1,100 @@
+  subject: string
+  action: string
+  eventTimestamp: number
+  attributes?: string // JSON serialized
```
any metadata on the event that is not in the schema already

rschlaefli commented on apps/audit/src/storage/entities.ts:
```diff
@@ -0,0 +1,100 @@
+  correlationClaims?: string // JSON serialized
+  stage?: string
+  schemaVersion: number
+  resource?: string
```
the resource that is targeted (e.g., live-quiz:XY, instance:XY)

rschlaefli commented on apps/audit/src/storage/entities.ts:
```diff
@@ -0,0 +1,100 @@
+  attributes?: string // JSON serialized
+  correlationId?: string
+  correlationClaims?: string // JSON serialized
+  stage?: string
```
not really used yet, but it might be useful for e.g. when we log at multiple stages in an async workflow

rschlaefli commented on apps/audit/src/storage/entities.ts:
```diff
@@ -0,0 +1,100 @@
+      : null,
+  })
+
+  return createHash('sha256').update(data).digest('hex').substring(0, 16)
```
we need to observe the performance impact of this

rschlaefli commented on apps/audit/src/storage/entities.ts:
```diff
@@ -0,0 +1,100 @@
+ * - Groups events by minute for efficient range queries
+ * - Uses shard (derived from eventId) to distribute load within a time bucket
+ */
+function generatePartitionKey(timestamp: number, eventId?: string): string {
```
we need to observe the performance characteristics of this

rschlaefli commented on apps/audit/src/app.ts:
```diff
@@ -0,0 +1,87 @@
+
+const app = new Hono()
+
+const customLoggerPrintFunc = (str: string, ...rest: string[]) => {
```
this connects hono (api framework) with pino (logging framework). all HTTP requests will then be logged nicely and in the same format as all of the other logs that we manually do in audit log

rschlaefli commented on apps/audit/src/config.ts:
```diff
@@ -0,0 +1,33 @@
+export type Config = z.infer<typeof configSchema>
+
+// Parse configuration once on module load to fail fast
+export const config = configSchema.parse(process.env)
```
this would fail the startup of the service if we have incorrectly configured env variables. would be a good idea to do this in all services gradually

rschlaefli commented on apps/frontend-pwa/src/components/liveQuiz/QuestionArea.tsx:
```diff
@@ -174,6 +196,16 @@ function QuestionArea({
       correlationKey,
     })
 
+    await auditLog.log({
```
first try on a UI generated audit log event

rschlaefli commented on apps/frontend-pwa/src/components/liveQuiz/QuestionArea.tsx:
```diff
@@ -57,6 +59,14 @@ function QuestionArea({
   const t = useTranslations()
 
+  const auditLog = useAuditClient({
+    assessmentMode: process.env.NEXT_PUBLIC_IS_ASSESSMENT === 'true',
```
right now these two are the same, as we have no audit log outside assessment mode

rschlaefli commented on apps/hatchet-worker-general/src/index.ts:
```diff
@@ -107,6 +109,14 @@ async function main() {
 
+  const auditClient = new AuditClient()
+
+  auditClient.log({
```
these are not working so well yet and we would probably need some unique identifier per worker/similar to identify them. but I think it might still be a good idea to track the "uptime" or crashes in the backend in the audit log? just for generally critical components though..

rschlaefli committed:
Remove bruno audit

rschlaefli committed:
global config in audit deploy

rschlaefli commented on packages/graphql/src/services/stacks.ts:
```diff
@@ -2617,256 +2617,362 @@ export async function respondToQuestion(
   },
```
Did not actually change much in this file beside adding comment for future logging and moving the entire thing inside try/catch (to be able to log the error later). could undo this change if we want to keep it assessment only for now

rschlaefli commented on packages/shared-components/src/hooks/useAuditClient.ts:
```diff
@@ -0,0 +1,113 @@
+      }
+
+      // All retries failed
+      setError(lastError)
```
we would need to decide what we do if even though it retried it still returns an error. probably let the user know with a toast? maybe its also just their internet that is down and the main requests fail anyways

rschlaefli commented on packages/util/src/audit.ts:
```diff
@@ -0,0 +1,145 @@
+        httpsAgent:
+          process.env.NODE_ENV === 'production'
+            ? undefined
+            : new https.Agent({ rejectUnauthorized: false }),
```
this needs to be done for dev mode as otherwise the https validation with our self made certificate would fail in certain cases

rschlaefli commented on util/wait-for-services.sh:
```diff
@@ -0,0 +1,394 @@
+#!/bin/bash
```
copied and extended this script from the cypress workflows -> it now waits as part of run app dependencies until the db etc. are up. could also migrate cypress workflows etc. to use this script from util/

rschlaefli commented on .versionrc.js:
```diff
@@ -1,46 +1,46 @@
+    //   filename: `deploy/charts/klicker-uzh-v2/Chart.yaml`,
+    //   updater: 'util/yaml-updater.js',
+    // },
+    ...packageJsonFiles.map((filename) => ({ filename, type: 'json' })),
```
much easier to parse the folders rather than having to hardcode all services -> so many were forgotten all the time and then not updated in package.json :D

rschlaefli commented on _run_app_dependencies.sh:
```diff
@@ -107,6 +107,12 @@ else
 	echo "Skipping pnpm run build"
 fi
 
+SERVICE_ENDPOINTS="" ./util/wait-for-services.sh || {
```
empty service endpoints means dependencies only

rschlaefli commented on apps/audit/test/api.test.ts:
```diff
@@ -0,0 +1,284 @@
+import { describe, expect, it } from 'vitest'
```
tests all work but we should have a real use case (right now we test with random data that it arrives, is processed, etc.) -> it might be interesting if we can run the assessment mode in cypress and validate the audit log output to be correct after the cypress run (maybe dump it to a file there). otherwise, the tests here are not so useful beside validating that data can be written and the api responds correctly (also why it is irrelevant what kind of data we store)

sjschlapbach commented on apps/audit/src/storage/entities.ts:
```diff
@@ -0,0 +1,100 @@
+    action: event.action,
+    timestamp: event.timestamp,
+    resource: event.resource || null,
+    // Include subset of attributes for uniqueness (avoid full object for size)
```
@rschlaefli  Do I understand this correctly that we do not take a subset of the attributes (contrary to what the comment states), but actually just cut off part of the hash we generate from the attributes? Does this make sense? This seems like we go through the potentially more expensive computation to compute the hash for the large object, throw away half of what is already a compact representation, but do not care about the other fields above, which have up to 200-400 characters according to the zod validation?

sjschlapbach commented on apps/frontend-pwa/src/components/liveQuiz/QuestionArea.tsx:
```diff
@@ -174,6 +196,16 @@ function QuestionArea({
+      action: AuditAction.PARTICIPANT_SUBMIT_RESPONSE,
+      resource: `instance:${instanceId}`,
+      scope: AuditScope.PUBLIC,
+      correlationId: correlationKey ?? undefined,
```
Maybe this is important: This `correlationKey` is not the same as the `correlationId` we use later on from within the response api through the assessment processor to track a response - the latter also contains the participant id, while this correlation key only contains the quiz id, instance id and execution ix in a format that can be validated

sjschlapbach commented on apps/response-api/src/index.ts:
```diff
@@ -187,11 +210,19 @@ async function handleAddAssessmentResponse(
+      'create-audit-log-entry',
+      {
+        subject: `participant:unknown`,
+        correlationId: correlationKey,
```
Same note as for the frontend use - the `correlationKey` is not the same as the `correlationId` we combine from the correlation key and the participant id further down (the same holds for all the following audit log entries?)

sjschlapbach commented on apps/response-api/src/index.ts:
```diff
@@ -275,12 +349,20 @@ async function handleAddAssessmentResponse(
+    {
+      scope: AuditScope.INTERNAL,
+      subject: `participant:${user.sub}`,
+      correlationId: correlationKey,
```
(same as above) - from here onwards, the correlationId is available and should be used

rschlaefli committed:
update claude model to sonnet 4.5

rschlaefli committed:
improve typesafety

rschlaefli committed:
type improvements

rschlaefli committed:
Merge branch 'v3' of https://github.com/uzh-bf/klicker-uzh into audit-log

# Conflicts:
#	packages/graphql/test/docker/docker-compose.test.yml
#	packages/graphql/test/helpers.ts

rschlaefli committed:
fix tests

rschlaefli committed:
Merge branch 'v3' of https://github.com/uzh-bf/klicker-uzh into audit-log

# Conflicts:
#	.github/workflows/test-graphql.yml

rschlaefli committed:
docs(project): add production-readiness review for assessment audit log

Detailed review of PR #4872 with evidence (file:line), findings on
auth/immutability/coverage/reliability, and an ordered roadmap toward
production readiness for the assessment-mode audit log.
